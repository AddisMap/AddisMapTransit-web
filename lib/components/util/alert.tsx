import React, { ReactElement } from 'react'
import styled from 'styled-components'

interface Props {
  children: React.ReactNode
}

export default function Alert({ children }: Props): ReactElement {
  const TripDescription = styled.div`
    background-color: rgba(19, 193, 193, 0.2);
    border-radius: 5px;
    box-shadow: 2px 2px 5px 1px rgba(0, 0, 0, 0.02);
    font-size: 12px;
    margin-bottom: 1em;
    padding: 0.75em 1em;
  `

  return <TripDescription>{children}</TripDescription>
}
