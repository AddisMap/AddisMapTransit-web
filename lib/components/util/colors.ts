/** A red color with enough contrast over white backgrounds */
export const RED_ON_WHITE = '#d1332e'

/** A gray color with enough contrast on white backgrounds. */
export const GRAY_ON_WHITE = '#757575'
