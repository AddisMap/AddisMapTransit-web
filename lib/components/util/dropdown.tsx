import { useIntl } from 'react-intl'
import React, { HTMLAttributes, useEffect, useRef, useState } from 'react'
import styled from 'styled-components'

interface Props extends HTMLAttributes<HTMLElement> {
  children: React.ReactNode
  label?: string
  name: JSX.Element
  pullRight?: boolean
}

const DropdownButton = styled.a`
  border: none;
  color: inherit;
  display: block;
  transition: all 0.1s ease-in-out;

  &:hover {
    background: rgba(0, 0, 0, 0.05) !important;
    cursor: pointer;
  }
  &.active {
    background: rgba(0, 0, 0, 0.1) !important;
  }
`
const DropdownMenu = styled.ul`
  background-clip: padding-box;
  background-color: #fff;
  border-radius: 4px;
  border: 1px solid rgba(0, 0, 0, 0.15);
  box-shadow: 0 6px 12px rgba(0, 0, 0, 0.175);
  color: #111;
  cursor: default;
  float: left;
  list-style: none;
  margin: 2px 0 0;
  min-width: 160px;
  padding: 5px 0;
  position: absolute;
  right: 0;
  text-align: left;
  top: 100%;
  z-index: 1000;

  hr {
    margin: 0;
    padding: 0;
  }
  li {
    cursor: pointer;
    padding: 5px 15px;
  }
  li.header {
    cursor: default;
  }
  li:not(.header):hover {
    background: rgba(0, 0, 0, 0.1);
  }
`
const DropdownContainer = styled.li``

/**
 * Renders a dropdown menu. By default, only a passed "name" is rendered. If clicked,
 * a floating div is rendered below the "name" with list contents inside. Clicking anywhere
 * outside the floating div will close the dropdown.
 */
const Dropdown = ({
  children,
  id,
  label,
  name,
  pullRight,
  style
}: Props): JSX.Element => {
  const [open, setOpen] = useState(false)
  const containerRef = useRef(null)
  useEffect(() => {
    // TODO: TYPE
    const handleClick = (e: any) => {
      if (
        !!containerRef.current &&
        // @ts-expect-error Typescript doesn't like this check
        !(containerRef?.current).contains(e.target)
      ) {
        setOpen(false)
      }
    }
    document.addEventListener('mousedown', handleClick)
    return () => {
      document.removeEventListener('mousedown', handleClick)
    }
  }, [containerRef])

  return (
    <DropdownContainer
      id={`${id}-wrapper`}
      ref={containerRef}
      role="presentation"
      style={{ float: pullRight ? 'right' : 'left' }}
    >
      <DropdownButton
        aria-controls={id}
        aria-expanded={open}
        aria-haspopup="listbox"
        aria-label={label}
        className={`${open && 'active'}`}
        id={`${id}-label`}
        onClick={() => setOpen(!open)}
        style={style}
        tabIndex={0}
      >
        {name}
        <span className="caret" style={{ marginLeft: 5 }} />
      </DropdownButton>
      {open && (
        <DropdownMenu
          aria-labelledby={`${id}-label`}
          id={id}
          role="listbox"
          tabIndex={-1}
        >
          {children}
        </DropdownMenu>
      )}
    </DropdownContainer>
  )
}

export default Dropdown
