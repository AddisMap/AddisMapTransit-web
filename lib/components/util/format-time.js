import moment from 'moment-timezone'
import { FormattedMessage } from 'react-intl'

export function FormattedTime ({endTime, startTime, timeFormat}) {
  return (
    <FormattedMessage
      id='common.time.departureArrivalTimes'
      values={{
        endTime: moment(endTime).format(timeFormat),
        startTime: moment(startTime).format(timeFormat)
      }}
    />
  )
}
