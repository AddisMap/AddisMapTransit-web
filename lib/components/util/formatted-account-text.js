import { FormattedMessage } from 'react-intl'
import PropTypes from 'prop-types'

/**
 * Returns a FormattedMessage component for the account wizard such that i18n IDs
 * are hardcoded and can be kept track of by format.js tools
 */
export default function FormattedAccountText ({key}) {
  switch (key) {
    case 'finish':
      return <FormattedMessage id='common.NewAccountWizard.finish' />
    case 'notifications':
      return <FormattedMessage id='common.NewAccountWizard.notifications' />
    case 'places':
      return <FormattedMessage id='common.NewAccountWizard.places' />
    case 'terms':
      return <FormattedMessage id='common.NewAccountWizard.terms' />
    case 'verify':
      return <FormattedMessage id='common.NewAccountWizard.verify' />
    default:
      return null
  }
}

FormattedAccountText.defaultProps = {
  key: ''
}

FormattedAccountText.propTypes = {
  key: PropTypes.string.isRequired
}
