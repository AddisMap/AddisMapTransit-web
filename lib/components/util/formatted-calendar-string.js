import { FormattedDate, FormattedMessage } from 'react-intl'
import coreUtils from '@opentripplanner/core-utils'
import moment from 'moment'
import PropTypes from 'prop-types'
import React from 'react'

import FormattedDayOfWeek from './formatted-day-of-week'

const { OTP_API_DATE_FORMAT } = coreUtils.time

/**
 * Returns a FormattedMessage component for date time preview calendar strings such that i18n IDs
 * are hardcoded and can be kept track of by format.js CLI tools
 */
const FormattedCalendarString = ({ date }) => {
  const previewDate = moment(date, OTP_API_DATE_FORMAT)
  const today = moment().startOf('day')
  const compareDate = previewDate.clone().startOf('day')
  const days = compareDate.diff(today, 'days')

  // Mimic moment.calendar logic (translation strings must come from language files)
  if (days > -7 && days < -1) {
    const dayId = previewDate.format('dddd').toLowerCase()
    const formattedDayOfWeek = <FormattedDayOfWeek day={dayId} />
    return (
      <FormattedMessage
        id="components.DateTimePreview.lastWeek"
        values={{ formattedDayOfWeek }}
      />
    )
  } else if (days === -1) {
    return <FormattedMessage id="common.dateExpressions.yesterday" />
  } else if (days === 0) {
    return <FormattedMessage id="common.dateExpressions.today" />
  } else if (days === 1) {
    return <FormattedMessage id="common.dateExpressions.tomorrow" />
  } else if (days > 1 && days < 7) {
    const dayId = previewDate.format('dddd').toLowerCase()
    return <FormattedDayOfWeek day={dayId} />
  } else {
    return <FormattedDate value={previewDate.valueOf()} />
  }
}

FormattedCalendarString.propTypes = {
  date: PropTypes.string
}

export default FormattedCalendarString
