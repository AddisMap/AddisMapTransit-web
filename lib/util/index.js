import * as itinerary from './itinerary'
import * as state from './state'

const OtpUtils = {
  itinerary,
  state
}

export default OtpUtils
