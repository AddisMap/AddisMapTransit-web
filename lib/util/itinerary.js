import { latLngBounds } from 'leaflet'
import coreUtils from '@opentripplanner/core-utils'
import moment from 'moment'

import { WEEKDAYS, WEEKEND_DAYS } from './monitored-trip'

export function getLeafletItineraryBounds(itinerary) {
  return latLngBounds(coreUtils.itinerary.getItineraryBounds(itinerary))
}

/**
 * Return a leaflet LatLngBounds object that encloses the given leg's geometry.
 */
export function getLeafletLegBounds(leg) {
  return latLngBounds(coreUtils.itinerary.getLegBounds(leg))
}

export function isBatchRoutingEnabled(otpConfig) {
  return Boolean(otpConfig.routingTypes.find((t) => t.key === 'BATCH'))
}

/**
 * Determines whether the specified Itinerary can be monitored.
 * @returns true if at least one Leg of the specified Itinerary is a transit leg,
 *   and none of the legs is a rental or ride hail leg (e.g. CAR_RENT, CAR_HAIL, BICYCLE_RENT, etc.).
 *   (We use the corresponding fields returned by OTP to get transit legs and rental/ride hail legs.)
 */
export function itineraryCanBeMonitored(itinerary) {
  let hasTransit = false
  let hasRentalOrRideHail = false

  if (itinerary && itinerary.legs) {
    for (const leg of itinerary.legs) {
      if (leg.transitLeg) {
        hasTransit = true
      }
      if (
        leg.rentedBike ||
        leg.rentedCar ||
        leg.rentedVehicle ||
        leg.hailedCar
      ) {
        hasRentalOrRideHail = true
      }
    }
  }

  return hasTransit && !hasRentalOrRideHail
}

export function getMinutesUntilItineraryStart(itinerary) {
  return moment(itinerary.startTime).diff(moment(), 'minutes')
}

/**
 * Gets the first transit leg of the given itinerary, or null if none found.
 */
function getFirstTransitLeg(itinerary) {
  return itinerary?.legs?.find((leg) => leg.transitLeg)
}

/**
 * Get the first stop ID from the itinerary in the underscore format required by
 * the startTransitStopId query param (e.g., TRIMET_12345 instead of TRIMET:12345).
 */
export function getFirstStopId(itinerary) {
  return getFirstTransitLeg(itinerary)?.from.stopId.replace(':', '_')
}

/**
 * Returns the set of monitored days that will be initially shown to the user
 * for the given itinerary.
 * @param itinerary The itinerary from which the default monitored days are extracted.
 * @returns ['monday' thru 'friday'] if itinerary happens on a weekday(*),
 *          ['saturday', 'sunday'] if itinerary happens on a saturday/sunday(*).
 * (*) For transit itineraries, the first transit leg is used to make
 * the determination. Otherwise, the itinerary startTime is used.
 */
export function getItineraryDefaultMonitoredDays(itinerary) {
  const firstTransitLeg = getFirstTransitLeg(itinerary)
  const startMoment = firstTransitLeg
    ? moment(firstTransitLeg.serviceDate, 'YYYYMMDD')
    : moment(itinerary.startTime)
  const dayOfWeek = startMoment.day()

  return dayOfWeek === 0 || dayOfWeek === 6 ? WEEKEND_DAYS : WEEKDAYS
}

export function itinerariesAreEqual(itinerary, other) {
  return itinerary.legs.every(
    (leg, index) =>
      other?.legs?.[index].from?.lat === leg.from?.lat &&
      other?.legs?.[index].to?.lat === leg.to?.lat &&
      other?.legs?.[index].from?.lon === leg.from?.lon &&
      other?.legs?.[index].to?.lon === leg.to?.lon
  )
}
