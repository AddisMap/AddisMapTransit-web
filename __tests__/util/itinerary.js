import {isTransit, makeRouteComparator} from '../../lib/util/itinerary'

const {
  route1,
  route2,
  route3,
  route4,
  route5,
  route6,
  route7,
  route8,
  route9,
  route10,
  route11,
  route12,
  route13
} = require('./__mocks__/itinerary.json')

function sortRoutes (...routes) {
  routes.sort(makeRouteComparator())
  return routes
}

describe('util > itinerary', () => {
  it('isTransit should work', () => {
    expect(isTransit('CAR')).toBeFalsy()
  })

  describe('routeComparator', () => {
    it('should sort routes based off of sortOrder', () => {
      expect(sortRoutes(route1, route2)).toMatchSnapshot()
    })

    it('should prioritize routes with valid sortOrder', () => {
      expect(sortRoutes(route2, route3)).toMatchSnapshot()
    })

    it('should sort routes based off of integer shortName with routes with same sortOrder', () => {
      expect(sortRoutes(route8, route2)).toMatchSnapshot()
    })

    it('should sort routes based off of integer shortName', () => {
      expect(sortRoutes(route3, route4)).toMatchSnapshot()
    })

    it('should prioritize routes with valid integer shortNames', () => {
      expect(sortRoutes(route4, route5)).toMatchSnapshot()
    })

    it('should sort routes based off of shortNames', () => {
      expect(sortRoutes(route5, route6)).toMatchSnapshot()
    })

    it('should prioritize routes with shortNames over those with just longNames', () => {
      expect(sortRoutes(route6, route7)).toMatchSnapshot()
    })

    it('should sort routes based off of longNames', () => {
      expect(sortRoutes(route9, route10)).toMatchSnapshot()
    })

    it('should sort routes on all of the criteria at once', () => {
      expect(sortRoutes(
        route1,
        route2,
        route3,
        route4,
        route5,
        route6,
        route7,
        route8,
        route9,
        route10,
        route11,
        route12,
        route13
      )).toMatchSnapshot()
    })

    it('should skip integer short name criteria when specified', () => {
      const routes = [route10, route11]
      routes.sort(makeRouteComparator(false))
      expect(routes).toMatchSnapshot()
    })

    it('should sort based off of route type', () => {
      const routes = [route12, route13]
      routes.sort(makeRouteComparator(false))
      expect(routes).toMatchSnapshot()
    })
  })
})
