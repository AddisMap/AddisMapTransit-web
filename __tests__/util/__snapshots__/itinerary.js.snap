// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`util > itinerary routeComparator should prioritize routes with shortNames over those with just longNames 1`] = `
Array [
  Object {
    "longName": "B-line",
    "shortName": "B",
    "sortOrder": -999,
  },
  Object {
    "longName": "A meandering route",
    "sortOrder": -999,
  },
]
`;

exports[`util > itinerary routeComparator should prioritize routes with valid integer shortNames 1`] = `
Array [
  Object {
    "longName": "Loop route",
    "shortName": "2",
    "sortOrder": -999,
  },
  Object {
    "longName": "A-line",
    "shortName": "A",
    "sortOrder": -999,
  },
]
`;

exports[`util > itinerary routeComparator should prioritize routes with valid sortOrder 1`] = `
Array [
  Object {
    "longName": "Around town",
    "shortName": "20",
    "sortOrder": 2,
  },
  Object {
    "longName": "Around another town",
    "shortName": "3",
    "sortOrder": -999,
  },
]
`;

exports[`util > itinerary routeComparator should sort routes based off of integer shortName 1`] = `
Array [
  Object {
    "longName": "Loop route",
    "shortName": "2",
    "sortOrder": -999,
  },
  Object {
    "longName": "Around another town",
    "shortName": "3",
    "sortOrder": -999,
  },
]
`;

exports[`util > itinerary routeComparator should sort routes based off of integer shortName with routes with same sortOrder 1`] = `
Array [
  Object {
    "longName": "Around town",
    "shortName": "20",
    "sortOrder": 2,
  },
  Object {
    "longName": "Zig-zagging route",
    "shortName": "30",
    "sortOrder": 2,
  },
]
`;

exports[`util > itinerary routeComparator should sort routes based off of longNames 1`] = `
Array [
  Object {
    "longName": "Express route",
    "shortName": "30",
    "sortOrder": 2,
  },
  Object {
    "longName": "Variation of express route",
    "shortName": "30",
    "sortOrder": 2,
  },
]
`;

exports[`util > itinerary routeComparator should sort routes based off of shortNames 1`] = `
Array [
  Object {
    "longName": "A-line",
    "shortName": "A",
    "sortOrder": -999,
  },
  Object {
    "longName": "B-line",
    "shortName": "B",
    "sortOrder": -999,
  },
]
`;

exports[`util > itinerary routeComparator should sort routes based off of sortOrder 1`] = `
Array [
  Object {
    "longName": "Around town",
    "shortName": "20",
    "sortOrder": 2,
  },
  Object {
    "longName": "Across town",
    "shortName": "10",
    "sortOrder": 10,
  },
]
`;

exports[`util > itinerary routeComparator should sort routes on all of the criteria at once 1`] = `
Array [
  Object {
    "longName": "Around town",
    "shortName": "20",
    "sortOrder": 2,
  },
  Object {
    "longName": "Express route",
    "shortName": "30",
    "sortOrder": 2,
  },
  Object {
    "longName": "Variation of express route",
    "shortName": "30",
    "sortOrder": 2,
  },
  Object {
    "longName": "Zig-zagging route",
    "shortName": "30",
    "sortOrder": 2,
  },
  Object {
    "longName": "Across town",
    "shortName": "10",
    "sortOrder": 10,
  },
  Object {
    "longName": "Loop route",
    "shortName": "2",
    "sortOrder": -999,
  },
  Object {
    "longName": "Around another town",
    "shortName": "3",
    "sortOrder": -999,
  },
  Object {
    "longName": "A-line",
    "shortName": "A",
    "sortOrder": -999,
  },
  Object {
    "longName": "B-line",
    "shortName": "B",
    "sortOrder": -999,
  },
  Object {
    "longName": "A meandering route",
    "sortOrder": -999,
  },
]
`;
